src/controllers/restful/PaymentController.ts

```diff
@@ -0,0 +1,204 @@
+export async function activatePaymentToPagoPa(
+  req: express.Request,
+  res: express.Response,
+  pagoPaConfig: PagoPaConfig
+): Promise<Either<ControllerError, PaymentsActivationResponse>> {
+  // Validate input
+  const errorOrPaymentsActivationRequest = PaymentsActivationRequest.decode(
+    req.params
+  );
+  if (errorOrPaymentsActivationRequest.isLeft()) {
+    return left(
+      RestfulUtils.sendErrorResponse(
+        res,
+        ControllerError.ERROR_INVALID_INPUT,
+        HttpErrorStatusCode.BAD_REQUEST
+      )
+    );
+  }
+
+  // Convert controller request to PagoPa request
+  const errorOrPaymentsActivationRequestPagoPa = PaymentsConverter.getPaymentsActivationRequestPagoPa(
+    pagoPaConfig,
+    errorOrPaymentsActivationRequest.value
+  );
+  if (errorOrPaymentsActivationRequestPagoPa.isLeft()) {
+    return left(
+      RestfulUtils.sendErrorResponse(
+        res,
+        ControllerError.ERROR_INVALID_INPUT,
+        HttpErrorStatusCode.BAD_REQUEST
+      )
+    );
+  }
+
+  // Require payment activation to PagoPa API
+  const errorOrPaymentActivationPagoPaResponse = await PaymentsService.sendPaymentsActivationRequestToPagoPaAPI(
+    errorOrPaymentsActivationRequestPagoPa.value,
+    pagoPaConfig
+  );
+
+  // Provide a response to applicant
+  if (errorOrPaymentActivationPagoPaResponse.isLeft()) {
+    RestfulUtils.sendUnavailableAPIError(res);
+    return left(ControllerError.ERROR_API_UNAVAILABLE);
+  }
+
+  // Check if request was rejected
+  if (
+    errorOrPaymentActivationPagoPaResponse.value.nodoAttivaRPTRisposta.esito ===
+    PPTPortTypes.Esito.KO
+  ) {
+    return left(
+      RestfulUtils.sendErrorResponse(
+        res,
+        ControllerError.REQUEST_REJECTED,
+        HttpErrorStatusCode.BAD_REQUEST
+      )
+    );
+  }
+
+  // Convert PagoPa response to controller response
+  const errorOrPaymentActivationResponse = PaymentsConverter.getPaymentsActivationResponse(
+    errorOrPaymentActivationPagoPaResponse.value
+  );
+  if (errorOrPaymentActivationResponse.isLeft()) {
+    return left(
+      RestfulUtils.sendErrorResponse(
+        res,
+        ControllerError.ERROR_INVALID_API_RESPONSE,
+        HttpErrorStatusCode.INTERNAL_ERROR
+      )
+    );
+  }
+  RestfulUtils.sendSuccessResponse(res, errorOrPaymentActivationResponse.value);
+  return right(errorOrPaymentActivationResponse.value);
+}
+
+/** Receive an async activation result frop PagoPA
+ * TODO: [#157910857] Creazione dei controller SOAP per l'esposizione dei servizi verso PagoPa
+ * TODO: [#158176380] Gestione della conferma di attivazione di un pagamento
+ */
+// tslint:disable-next-line:no-empty
+export async function notifyPaymentStatusToAPINotifica(): Promise<void> {}
+
+// Generate a Session Token to follow a stream of requests
+function generateCodiceContestoPagamento(): Either<
+  Error,
+  CodiceContestoPagamento
+> {
+  return CodiceContestoPagamento.decode(uuid.v1()).mapLeft(() => {
```

Comment:
when does this decoding fail? this looks like one of the very few cases where we can safely cast